package heron.proto.stmgr;
option java_package = "com.twitter.heron.proto.stmgr";
option java_outer_classname = "StreamManager";

import "common.proto";
import "stats.proto";
import "topology.proto";
import "physical_plan.proto";
import "tuple.proto";

//
// Messages sent by topologymaster
//

message NewPhysicalPlanMessage {
  required heron.proto.system.PhysicalPlan new_pplan = 1;
}

//
// Request/Responses by the instance
//
message RegisterInstanceRequest {
  // Information about the instance registering with us
  required heron.proto.system.Instance instance = 1;
  required string topology_name = 2;
  required string topology_id = 3;
}

message RegisterInstanceResponse {
  required heron.proto.system.Status status = 1;
  // If the assignment is known, send it
  optional heron.proto.system.PhysicalPlan pplan = 2;
}

message NewInstanceAssignmentMessage {
  required heron.proto.system.PhysicalPlan pplan = 1;
}

// Tuples exchanged between stream managers and instances
message TupleMessage {
  required heron.proto.system.HeronTupleSet set = 1;
}

// This is the message that the instance sends
// when it wants stmgr to store its state
message InstanceStateCheckpoint {
  required string checkpoint_id = 1;
  required bytes state = 2;
}

// This is the message that the stmgr sends to its
// local spouts to begin initiating checkpoint
message InitiateStatefulCheckpoint {
  required string checkpoint_id = 1;
}

// messages exchanged between stream managers
message StrMgrHelloRequest {
  required string topology_name = 1;
  required string topology_id = 2;
  required string stmgr = 3;
}

message StrMgrHelloResponse {
  required heron.proto.system.Status status = 1;
}

// Tuples exchanged between stream managers
message TupleStreamMessage {
  required int32 task_id = 1;
  required heron.proto.system.HeronTupleSet set = 2;
}

message StartBackPressureMessage {
  required string topology_name = 1;
  required string topology_id = 2;
  required string stmgr = 3;
  required string message_id = 4;
}

message StopBackPressureMessage {
  required string topology_name = 1;
  required string topology_id = 2;
  required string stmgr = 3;
  required string message_id = 4;
}

// Tuples exchanged between stream managers and instances
message TupleMessage2 {
  required heron.proto.system.HeronTupleSet2 set = 1;
}

// Tuples exchanged between stream managers
message TupleStreamMessage2 {
  required int32 task_id = 1;
  // serialized data
  required bytes set = 2;
}

// Messages sent from/to Checkpoint Manager

// This is the message that StMgr sends to Checkpoint Mgr
// to store instance checkpoint data
message SaveStateCheckpoint {
  // Information about the instance whose state this is
  required heron.proto.system.Instance instance = 1;
  required InstanceStateCheckpoint checkpoint = 2;
}
